### PR TITLE
[CALCITE-4982] NonNull field shouldn't be pushed down into leaf of outer-join in 'ProjectJoinTransposeRule'

### PR DESCRIPTION
**Pull request by wojustme**
This pr is prepared for [ISSUE-4982](https://issues.apache.org/jira/browse/CALCITE-4982)

### CODE REVIEW

File: core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml
Comment (NobiGo) [inline core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:9032]: @wojustme If the $3 is NULL, then what will happen in `EXPR$1=[CAST($3):VARCHAR(10) NOT NULL])`? Because the EXPR1 will execute any time. right? 
But  the plan bafore, when we confirm $3 is not NULL, we will to calcute `[CAST($3):VARCHAR(10) NOT NULL])`,
So please try when job include NULL value not just claim this filed can NULL, Plase check the data, I don't know, Just want to confirm.

**Review by NobiGo — CHANGES_REQUESTED** [review]

File: core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml
Comment (wojustme) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:9032]: @NobiGo 
It's good question for me. 
In my project, we use calcite as an sql-proxy, and send dialectal sql to various query engine.
`EXPR$1=[CAST($3):VARCHAR(10) NOT NULL])` will be translated to sql fragment of `CAST(job AS VARCHAR)`
So, I didn't pay attention to this problem. I try to find some codes or docs to explain this question.

In calcite's linq4j, `CAST($3):VARCHAR(10) NOT NULL` will be translated to `input_isNull0 ? (String) null : input_value0`

Comment (NobiGo) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:9031]: Hi, Please try validate this SQL result in out.iq and submit the test in this PR? I don't know.

**Comment by wojustme** [comment]
Hi @NobiGo 
Take your time to review this conversation.
Thanks.

**Comment by wojustme** [comment]
Hi @NobiGo 
Are you going on reviewing this pr?

**Comment by libenchao** [comment]
What's the status of this PR?

**Comment by wojustme** [comment]
@libenchao 
Thank you for reminding me, and I almost forgot this issue.
I will update this pr as soon as possible.

**Comment by wojustme** [comment]
@libenchao 
This issue/pr was so long that it took me a little time to review it.
Two solutions:
1. Avoid pushing down with the operand of coalesce's operator
2. Keep pushing down, and fix `org.apache.calcite.plan.Strong#policy(org.apache.calcite.rex.RexNode)`

Let's discuss in the [JIRA](https://issues.apache.org/jira/browse/CALCITE-4982)

**Comment by wojustme** [comment]
Add default config: PreserveExprCondition
Shouldn't push non-null field, which is wrapped by CAST's operator.
Hi @libenchao 
Please review this pr again, thanks a lot.

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java
Comment (libenchao) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java:163]:
```suggestion
          // Do not push down over's expression by default
```

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java
Comment (libenchao) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java:172]: casting `not null` to `nullable` could be pushed down?

**Review by libenchao — COMMENTED** [review]
@wojustme Thanks for updating. 
In the JIRA discussion, I was suggesting not pushing `CASE`'s operands partially through `Join`, which is a more general way. However, it seems your approach could also fix this issue.  

I'm fine with current approach since it indeed works for this issue, and leave others to later issues, maybe in CALCITE-5315.

**Comment by wojustme** [comment]
In my opinion, this case only works on in CALCITE's optimizer or validator, which some expresion only change nullable attribute.
we couldn't write an sql `cast(c1 as not integer)`, and we only write an sql `cast(c1 as integer)`.

This pr will fix the issue with a little modification. Maybe we get a good solution from the discussion of [CALCITE-5315](https://issues.apache.org/jira/browse/CALCITE-5315).

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java
Comment (libenchao) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java:163]: kindly ping.

**Comment by wojustme** [comment]
Hi @libenchao 
PR is ready, and could you take your time to review it again, thanks.

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java
Comment (libenchao) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java:163]: @wojustme Here is one missing comment which hasn't been addressed.

Comment (wojustme) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java:163]: @libenchao Thanks for your reminder, but I check this pr, and I found that I have update this code, screenshot as follows:
image

Comment (libenchao) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java:163]: It's in different line, I'm saying line 163

Comment (wojustme) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java]: I'm sorry for it.
Please review it, thanks a lot.
@libenchao

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java
Comment (libenchao) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java:167]: move this into to `if` branch

**Review by libenchao — COMMENTED** [review]
@wojustme LGTM with one minor comment, after fixing that, I'll merge this.

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java
Comment (wojustme) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java]: Hi @libenchao 
I have moved this code into `else` branch, I guess this's your suggestion.

Comment (libenchao) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java]: Sorry that I was not clear enough on this. I mean:
```java
          if (expr instanceof RexOver) {
            return false;
          }
          if (SqlKind.CAST == expr.getKind()) {
            final RelDataType relType = expr.getType();
            final RexCall castCall = (RexCall) expr;
            final RelDataType operand0Type = castCall.getOperands().get(0).getType();
            if (relType.getSqlTypeName() == operand0Type.getSqlTypeName()
                && operand0Type.isNullable() && !relType.isNullable()) {
              // Do not push down not nullable cast's expression with the same type by default
              // eg: CAST($1):VARCHAR(10) NOT NULL, and type of $1 is nullable VARCHAR(10)
              return false;
            }
          }
```

Comment (wojustme) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java]: Got it, and pr has been updated.
Please review it.
@libenchao

**Review by libenchao — APPROVED** [review]
+1, merging.

**Comment by libenchao** [comment]
fixed via https://github.com/apache/calcite/commit/43db14cf8e08230c309eeb425a568e77db964a46

**Pull request closed without merging**
